src/Boson_Sampling_Utilities.py

```diff
@@ -10,17 +23,100 @@ def per(mtx, column, selected, prod):
     """
     if column == mtx.shape[1]:
         return prod
-    else:
-        result = 0
-        for row in range(mtx.shape[0]):
-            if row not in selected:
-                result = result \
-                         + per(mtx, column + 1, selected + [row], prod * mtx[row, column])
-        return result
 
+    result = 0
+    for row in range(mtx.shape[0]):
+        if row not in selected:
+            result = result \
+                     + permanent_recursive_part(mtx, column + 1, selected + [row], prod * mtx[row, column])
+    return result
+
+
+def particle_state_to_modes_state(particle_state: ndarray, observed_modes_number: int) -> ndarray:
+    modes_state = zeros(observed_modes_number)
 
-def calculate_permanent(mat):
+    # Adding the particle to it's mode.
+    for particle in particle_state:
+        modes_state[int(particle)] += 1
+
+    # numbers of particles in each mode
+    return modes_state
+
+
+def modes_state_to_particle_state(mode_state: ndarray, particles_number: int) -> ndarray:
     """
-    Returns the permanent of the matrix mat.
+        Return given mode-basis state in particle basis.
+
+        :param mode_state: Input state in mode-basis.
+        :param particles_number: Number of particles.
+        :return: Given mode-basis state in particle basis.
     """
-    return per(mat, 0, [], 1)
+
+    number_of_observed_modes = len(mode_state)
+    modes = mode_state.copy()
+    particles_state = zeros(particles_number)
+
+    i = k = 0
+    while i < number_of_observed_modes:
+
+        if modes[i] > 0:
+            modes[i] -= 1
+            particles_state[k] = int(i)
+            k += 1
+        else:
+            i += 1
+
+    return particles_state
+
+
+def generate_possible_outputs(number_of_particles: int, number_of_modes: int) -> List[ndarray]:
+    outputs = []
+
+    output = zeros(number_of_modes)
+    output[0] = number_of_particles
+    outputs.append(output)
+
+    while outputs[-1][number_of_modes - 1] != number_of_particles:
```

Comment:
Can there be more particles than the number of particles? If not, checking for ` < number_of_particles` would be safer.

Reply:
+1